x/superfluid/keeper/migrate.go

```diff
@@ -1,99 +1,302 @@
+func (k Keeper) migrateNonSuperfluidLockBalancerToConcentrated(ctx sdk.Context, sender sdk.AccAddress, poolIdLeaving uint64, preMigrationLock *lockuptypes.PeriodLock, gammSharesInLock sdk.Coin, sharesToMigrate sdk.Coin, concentratedPool cltypes.ConcentratedPoolExtension, remainingLockTime time.Duration) (positionId uint64, amount0, amount1 sdk.Int, liquidity sdk.Dec, joinTime time.Time, gammLockId, concentratedLockId uint64, err error) {
+	// Save unlocking state of lock before force unlocking
+	wasUnlocking := preMigrationLock.IsUnlocking()
+
+	// Force unlock, validate the provided sharesToMigrate, and exit the balancer pool.
+	// This will return the coins that will be used to create the concentrated liquidity position.
+	exitCoins, err := k.validateSharesToMigrateUnlockAndExitBalancerPool(ctx, sender, poolIdLeaving, preMigrationLock, gammSharesInLock, sharesToMigrate)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+	}
+
+	// Create a new lock that is unlocking for the remaining time of the old lock.
+	// Regardless of the previous lock's status, we create a new lock that is unlocking.
+	positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, err = k.clk.CreateFullRangePositionUnlocking(ctx, concentratedPool, sender, exitCoins, remainingLockTime)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+	}
+
+	// If there are remaining gamm shares, we must re-lock them.
+	remainingGammShares := gammSharesInLock.Sub(sharesToMigrate)
+	if !remainingGammShares.IsZero() {
+		// Create a new lock with the remaining gamm shares for the remaining lock time.
+		newLock, err := k.lk.CreateLock(ctx, sender, sdk.NewCoins(remainingGammShares), remainingLockTime)
+		gammLockId = newLock.ID
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+		}
+
+		// If the gamm lock was unlocking, we begin the unlock from where it left off.
+		if wasUnlocking {
+			_, err := k.lk.BeginForceUnlock(ctx, newLock.ID, newLock.Coins)
+			if err != nil {
+				return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+			}
+		}
+	}
+
+	return positionId, amount0, amount1, liquidity, joinTime, gammLockId, concentratedLockId, nil
+}
+
+// prepareMigration prepares for the migration of gamm LP tokens from the Balancer pool to the Concentrated pool. It performs the following steps:
+//
+// 1. Gets the pool ID of the Balancer pool from the gamm share denomination.
+// 2. Ensures a governance-sanctioned link exists between the Balancer pool and the Concentrated pool.
+// 3. Validates that the lock corresponds to the sender, contains the correct denomination of LP shares, and retrieves the gamm shares from the lock.
+// 4. Determines the remaining time on the lock.
+// 5. Checks if the lock has a corresponding synthetic lock, indicating it is superfluid delegated or undelegating.
```

Comment:
Please list the return values, there are quite a few. This will make it browsing through code easier in the future

Reply:
https://github.com/osmosis-labs/osmosis/pull/4885/commits/f2a52f14868a87ba3bbab6744141d75a65d36b04